File: scripts/analyze_airbnb.R
# AirBnB Data Analysis Script
# STAT301 Project
# 
# This script performs exploratory data analysis on AirBnB listings data

# Load required libraries
library(tidyverse)
library(ggplot2)

# Set working directory to project root
setwd(dirname(dirname(rstudioapi::getActiveDocumentContext()$path)))

# Load data
cat("Loading AirBnB data...\n")
airbnb_data <- read_csv("data/sample_data.csv")

# Display basic information
cat("\n=== Dataset Overview ===\n")
cat("Number of rows:", nrow(airbnb_data), "\n")
cat("Number of columns:", ncol(airbnb_data), "\n")
cat("\nColumn names:\n")
print(colnames(airbnb_data))

cat("\n=== Data Structure ===\n")
str(airbnb_data)

cat("\n=== Summary Statistics ===\n")
summary(airbnb_data)

# Price Analysis
cat("\n=== Price Analysis ===\n")
cat("Average price: $", mean(airbnb_data$price, na.rm = TRUE), "\n")
cat("Median price: $", median(airbnb_data$price, na.rm = TRUE), "\n")
cat("Price range: $", min(airbnb_data$price, na.rm = TRUE), " - $", 
    max(airbnb_data$price, na.rm = TRUE), "\n")

# Room type distribution
cat("\n=== Room Type Distribution ===\n")
room_type_counts <- table(airbnb_data$room_type)
print(room_type_counts)

# Neighbourhood analysis
cat("\n=== Top 5 Neighbourhoods by Number of Listings ===\n")
neighbourhood_counts <- sort(table(airbnb_data$neighbourhood), decreasing = TRUE)
print(head(neighbourhood_counts, 5))

# Create visualizations

# 1. Price distribution
cat("\nGenerating price distribution plot...\n")
p1 <- ggplot(airbnb_data, aes(x = price)) +
  geom_histogram(bins = 30, fill = "steelblue", color = "black", alpha = 0.7) +
  labs(title = "Distribution of AirBnB Listing Prices",
       x = "Price ($)",
       y = "Count") +
  theme_minimal()

ggsave("figures/price_distribution.png", p1, width = 8, height = 6, dpi = 300)

# 2. Price by room type
cat("Generating price by room type plot...\n")
p2 <- ggplot(airbnb_data, aes(x = room_type, y = price, fill = room_type)) +
  geom_boxplot() +
  labs(title = "Price Distribution by Room Type",
       x = "Room Type",
       y = "Price ($)") +
  theme_minimal() +
  theme(legend.position = "none")

ggsave("figures/price_by_room_type.png", p2, width = 8, height = 6, dpi = 300)

# 3. Reviews vs Price
cat("Generating reviews vs price scatter plot...\n")
p3 <- ggplot(airbnb_data, aes(x = number_of_reviews, y = price)) +
  geom_point(alpha = 0.6, color = "steelblue") +
  geom_smooth(method = "lm", color = "red", se = TRUE) +
  labs(title = "Relationship between Number of Reviews and Price",
       x = "Number of Reviews",
       y = "Price ($)") +
  theme_minimal()

ggsave("figures/reviews_vs_price.png", p3, width = 8, height = 6, dpi = 300)

# 4. Neighbourhood distribution
cat("Generating neighbourhood distribution plot...\n")
p4 <- airbnb_data %>%
  count(neighbourhood) %>%
  top_n(10, n) %>%
  ggplot(aes(x = reorder(neighbourhood, n), y = n)) +
  geom_col(fill = "steelblue") +
  coord_flip() +
  labs(title = "Top 10 Neighbourhoods by Number of Listings",
       x = "Neighbourhood",
       y = "Number of Listings") +
  theme_minimal()

ggsave("figures/neighbourhood_distribution.png", p4, width = 8, height = 6, dpi = 300)

# Statistical tests

# Test: Is there a significant difference in prices between room types?
cat("\n=== Statistical Test: Price Differences by Room Type ===\n")
cat("ANOVA test for price differences by room type:\n")
anova_result <- aov(price ~ room_type, data = airbnb_data)
print(summary(anova_result))

# Correlation analysis
cat("\n=== Correlation Analysis ===\n")
cat("Correlation between price and number of reviews:\n")
cor_test <- cor.test(airbnb_data$price, airbnb_data$number_of_reviews)
print(cor_test)

cat("\n=== Analysis Complete ===\n")
cat("Figures saved to: figures/\n")
cat("Check the report in: reports/\n")

File: scripts/analyze_airbnb.py
#!/usr/bin/env python3
"""
AirBnB Data Analysis Script
STAT301 Project

This script performs exploratory data analysis on AirBnB listings data
"""

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from scipy import stats
import os

# Set style for plots
sns.set_style("whitegrid")
plt.rcParams['figure.figsize'] = (10, 6)

def load_data(filepath):
    """Load AirBnB data from CSV file"""
    print(f"Loading AirBnB data from {filepath}...")
    df = pd.read_csv(filepath)
    print(f"Data loaded successfully! Shape: {df.shape}")
    return df

def basic_info(df):
    """Display basic information about the dataset"""
    print("\n=== Dataset Overview ===")
    print(f"Number of rows: {df.shape[0]}")
    print(f"Number of columns: {df.shape[1]}")
    print(f"\nColumn names:\n{df.columns.tolist()}")
    
    print("\n=== Data Types ===")
    print(df.dtypes)
    
    print("\n=== Missing Values ===")
    print(df.isnull().sum())
    
    print("\n=== Summary Statistics ===")
    print(df.describe())

def analyze_prices(df):
    """Analyze price distribution"""
    print("\n=== Price Analysis ===")
    print(f"Average price: ${df['price'].mean():.2f}")
    print(f"Median price: ${df['price'].median():.2f}")
    print(f"Price range: ${df['price'].min():.2f} - ${df['price'].max():.2f}")
    print(f"Standard deviation: ${df['price'].std():.2f}")

def analyze_room_types(df):
    """Analyze room type distribution"""
    print("\n=== Room Type Distribution ===")
    print(df['room_type'].value_counts())
    print("\n=== Average Price by Room Type ===")
    print(df.groupby('room_type')['price'].mean().sort_values(ascending=False))

def analyze_neighbourhoods(df):
    """Analyze neighbourhood distribution"""
    print("\n=== Top 10 Neighbourhoods by Number of Listings ===")
    print(df['neighbourhood'].value_counts().head(10))
    print("\n=== Average Price by Top 5 Neighbourhoods ===")
    top_neighbourhoods = df['neighbourhood'].value_counts().head(5).index
    print(df[df['neighbourhood'].isin(top_neighbourhoods)].groupby('neighbourhood')['price'].mean().sort_values(ascending=False))

def create_visualizations(df, output_dir='figures'):
    """Create and save visualizations"""
    os.makedirs(output_dir, exist_ok=True)
    
    # 1. Price distribution
    print("\nGenerating price distribution plot...")
    plt.figure(figsize=(10, 6))
    plt.hist(df['price'], bins=30, color='steelblue', alpha=0.7, edgecolor='black')
    plt.xlabel('Price ($)', fontsize=12)
    plt.ylabel('Count', fontsize=12)
    plt.title('Distribution of AirBnB Listing Prices', fontsize=14, fontweight='bold')
    plt.grid(axis='y', alpha=0.3)
    plt.tight_layout()
    plt.savefig(f'{output_dir}/price_distribution.png', dpi=300, bbox_inches='tight')
    plt.close()
    
    # 2. Price by room type
    print("Generating price by room type plot...")
    plt.figure(figsize=(10, 6))
    df.boxplot(column='price', by='room_type', patch_artist=True)
    plt.xlabel('Room Type', fontsize=12)
    plt.ylabel('Price ($)', fontsize=12)
    plt.title('Price Distribution by Room Type', fontsize=14, fontweight='bold')
    plt.suptitle('')  # Remove default title
    plt.tight_layout()
    plt.savefig(f'{output_dir}/price_by_room_type.png', dpi=300, bbox_inches='tight')
    plt.close()
    
    # 3. Reviews vs Price
    print("Generating reviews vs price scatter plot...")
    plt.figure(figsize=(10, 6))
    plt.scatter(df['number_of_reviews'], df['price'], alpha=0.6, color='steelblue')
    
    # Add regression line
    z = np.polyfit(df['number_of_reviews'], df['price'], 1)
    p = np.poly1d(z)
    plt.plot(df['number_of_reviews'].sort_values(), 
             p(df['number_of_reviews'].sort_values()), 
             "r--", alpha=0.8, linewidth=2)
    
    plt.xlabel('Number of Reviews', fontsize=12)
    plt.ylabel('Price ($)', fontsize=12)
    plt.title('Relationship between Number of Reviews and Price', fontsize=14, fontweight='bold')
    plt.grid(alpha=0.3)
    plt.tight_layout()
    plt.savefig(f'{output_dir}/reviews_vs_price.png', dpi=300, bbox_inches='tight')
    plt.close()
    
    # 4. Neighbourhood distribution
    print("Generating neighbourhood distribution plot...")
    plt.figure(figsize=(10, 8))
    top_neighbourhoods = df['neighbourhood'].value_counts().head(10)
    plt.barh(range(len(top_neighbourhoods)), top_neighbourhoods.values, color='steelblue')
    plt.yticks(range(len(top_neighbourhoods)), top_neighbourhoods.index)
    plt.xlabel('Number of Listings', fontsize=12)
    plt.ylabel('Neighbourhood', fontsize=12)
    plt.title('Top 10 Neighbourhoods by Number of Listings', fontsize=14, fontweight='bold')
    plt.grid(axis='x', alpha=0.3)
    plt.tight_layout()
    plt.savefig(f'{output_dir}/neighbourhood_distribution.png', dpi=300, bbox_inches='tight')
    plt.close()
    
    print(f"\nAll figures saved to: {output_dir}/")

def statistical_tests(df):
    """Perform statistical tests"""
    print("\n=== Statistical Tests ===")
    
    # ANOVA test: Price differences by room type
    print("\n1. ANOVA Test: Price Differences by Room Type")
    room_types = df['room_type'].unique()
    groups = [df[df['room_type'] == rt]['price'].values for rt in room_types]
    f_stat, p_value = stats.f_oneway(*groups)
    print(f"F-statistic: {f_stat:.4f}")
    print(f"P-value: {p_value:.4f}")
    if p_value < 0.05:
        print("Result: Significant difference in prices between room types (p < 0.05)")
    else:
        print("Result: No significant difference in prices between room types (p >= 0.05)")
    
    # Correlation test: Price vs Number of Reviews
    print("\n2. Correlation Test: Price vs Number of Reviews")
    corr_coef, p_value = stats.pearsonr(df['price'], df['number_of_reviews'])
    print(f"Correlation coefficient: {corr_coef:.4f}")
    print(f"P-value: {p_value:.4f}")
    if abs(corr_coef) < 0.3:
        strength = "weak"
    elif abs(corr_coef) < 0.7:
        strength = "moderate"
    else:
        strength = "strong"
    direction = "positive" if corr_coef > 0 else "negative"
    print(f"Result: {strength.capitalize()} {direction} correlation")

def main():
    """Main analysis function"""
    # Load data
    df = load_data('data/sample_data.csv')
    
    # Perform analyses
    basic_info(df)
    analyze_prices(df)
    analyze_room_types(df)
    analyze_neighbourhoods(df)
    
    # Create visualizations
    create_visualizations(df)
    
    # Statistical tests
    statistical_tests(df)
    
    print("\n=== Analysis Complete ===")
    print("Figures saved to: figures/")
    print("Check the report in: reports/")

if __name__ == "__main__":
    main()
